NIFI-5986 Adding "Stop & Configure" button functionality to Processor…

aeaversa opened the pull request:
NIFI-5986 Added "Stop & Configure" button functionality to Processor Details modal.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

aeaversa commented:
@mcgilman - Thanks Matt!  So glad you got a chance to take a look. So, the motivation behind the change was purely to improve navigational efficiency and response time for end users in quickly being able to identify and correct a started processor that somehow had an incorrect configuration; limiting the amount of flow files that can traverse through the misconfigured processor. In working with both novice and advanced users of NiFi for the last year, I've had consistent requests that there be a simple button that allows a user to directly stop a processor and automatically allow editing of the processor's properties. The current navigational route from viewing a started processor's details (read only)  to editing its properties in the configure processor modal was always reported as a bit cumbersome: (ie - user clicks to open the  processor details modal of a started processor and notices a properties change is necessary > user clicks to close the processor details modal > user right/left clicks to stop the processor > user clicks to open what looks like the same processor details modal dialog, but now allows them to edit the desired properties).  I tried to keep the changes as simple as possible when I sent the PR. In this change, the processor details modal should  only present the "Stop & Configure" button when its launched directly from the canvas interface to limit any conflicts with  where it is used on other interfaces (ie: Summary, etc etc). ... as well it will disable the button if the processor is in any state other then RUNNING or the user demonstrates insufficient permissions. If there are other concerns you have, please let me know what those are and I can work to get those addressed. My thanks again for taking a look!

Take care,

--alex

mosermw commented:
For what it's worth, I'm super in favor of user experience improvements like this.  When using NiFi regularly and sometimes all day, extra clicks get annoying for something you are doing constantly.  I've heard the same feedback from users myself, that this feature would be useful.

My only concern is getting this right for processors that do not stop instantly.  I get super annoyed when I stop a long running processor that's currently using threads and I have to Refresh the canvas waiting for it to stop.  It would be great for this feature to Stop the processor, wait for it to fully stop, then Configure.  There might need to be a timeout in there if the processor doesn't stop in 10 seconds or so.

aeaversa commented:
@mosermw - Thanks Mike! Yeah, I totally agree that there should be another option outside of "Terminate" that allows a processor to naturally stop on its own. I propose to add a "WAIT" button on the Terminate dialog versus "NO" and when clicked, simply poll the processor state until the processor has chewed through its active threads and fully stopped before ultimately presenting the Configure Processor modal. Let me know if this sounds like an optimal solution in your opinion to this issue.

mcgilman commented:
@aeaversa @mosermw Thanks for the comments. This definitely helps. I was just trying to be sure that we were focusing on the right problem. The request makes sense.

Your PR does actually address some of the concerns I had which I admittedly overlooked when reading the changeset on my phone. I can also get behind the proposed solution but I'm CCing a couple community members who have assisted with UX issues in the past to get their input on the matter as well.

I do still have a few concerns with the proposed code changes. I'll add them in the review and can discuss them more there. Thanks again!

CC @moranr @andrewmlim

aeaversa commented:
@mcgilman -  Thanks again for all the great feedback. I'll sit tight for input from the others and next week I'll work to refactor to the consensus. I have a working branch right now that does have the polling logic within the terminate dialog under a  "WAIT" button (in lieu of the "NO" button) and provides feedback to the user on the modal as to how many threads remain active. I would say that being able to stop and terminate immediately is highly desired by some of the users I've discussed the feature with, but if you think that has deleterious effects elsewhere in the application, I can take the terminate logic out entirely if needed. My thanks again for all your help with this.

mcgilman commented:
@aeaversa I don't think the terminate logic needs to be completely removed but it should only be used when necessary. Generally speaking, when possible the processor threads should be allowed to complete gracefully.

I could see possibly showing some sort of dialog/status if there were long-running threads and showing how many were remaining. Then maybe there provide an option there to cancel or terminate. But this is where I'd like to make sure that @moranr and @andrewmlim (and anyone else for that matter) can chime in if they want to.

Thanks!

moranr commented:
> I could see possibly showing some sort of dialog/status if there were long-running threads and showing how many were remaining. Then maybe there provide an option there to cancel or terminate.

Agreed @mcgilman -- if there are active threads, it would be nice to display messaging to the user (within the current dialog, new dialog, ...?) explaining that fact, along with some type of timer or indicator to let user know the application is watching/waiting to allow it to end gracefully, at which point it will open the configurable settings. Then, as mentioned, provide the option to force the termination immediately (as it may be the only viable option).

Also thinking it'd be nice if the user chooses to wait it out, to close that dialog and continue to work in the app. After a certain amount of time, reopen the dialog to remind them we're still waiting and again provide the terminate option. 

Thanks for your work on this @aeaversa . I will plan on testing this out later and provide more detailed feedback.

aeaversa commented:
@moranr - Yeah I actually like this idea a lot. How about this... we present the processor configuration dialog, with its "Apply" button disabled, as soon as they hit the "Stop & Configure" button on the processor details dialog, thereby allowing the user to begin to view/change properties while the threads continue to closeout normally with a subtle feedback mechanism on the configuration dialog that displays the active thread count to the user as the processor is polled (i.e. "Stopping..3 active threads remain"). From there we can simply present a "Terminate" button to the bottom of the dialog, should the user want to elect to do so. Once the process reaches a zero thread count,  we can simply "enable" the Apply button and remove the "Terminate" button and the form would function as originally intended. Let me know if this design seems like a more natural flow to you guys and my thanks again for all your help with this!

aeaversa commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/nf-processor-details.js:
```diff
@@ -274,6 +280,82 @@
+                                         processorResponse.operatePermissions.canWrite) );
+                         },
+                         handler: {
+                             click: function () {
```
@mcgilman - 100% agree with this approach and will refactor accordingly.

aeaversa commented:
@moranr - Thanks Rob!  Should have a fresh commit out on the PR in a few days and look forward to hearing your suggestions when you can take a test drive.

mcgilman commented:
@aeaversa @moranr I'm definitely not opposed to the idea's here. Just wanted to offer a few thoughts.

Currently, we cannot open the Configure dialog while there are still active threads. This path forward moves away from that. We'd likely want to update the Configure context menu availability as well. The one benefit the current solution offers, however, is that if the processor is reporting any issues they are available via its bulletins. If we already have the Configuration dialog open there would be no way to see these. This is something that I hadn't even considered during my previous comments. We could also make these bulletins available, in addition to the active threads, but I wonder if it starts to get too muddled all in the Configuration dialog.

Not sure, maybe it's not necessary. Thoughts? 

Also looking forward to checking it out.

aeaversa commented:
@mcgilman - I definitely see your point. It really would be nice to have the bulletins available for the processor on the dialog. I'll look into it as I refactor.

aeaversa commented:
@mcgilman @moranr - I went ahead and updated the code to allow the Processor Configuration/Details dialog(s) to support the display of threads and bulletins in the same style as is displayed on the processors on the canvas. A terminate prompt is on the Processor Configuration dialog "Apply" button as well, such that should the end user hit "Apply" and threads are still active, it will simply provide them an option to terminate or wait for the threads to finish normally. I believe this is a good solution to allow threads to naturally close out in the background as the user is making their changes. I did not institute an additional polling process to inspect the thread state, but rather simply synced the thread state value off the current canvas value for the component, since that interface already has a polling process active. Additionally, per @mcgilman suggestion, I added the Stop & Configure option to the context menu as well. Please take a test drive and let me know your feedback whenever you guys get a chance. My thanks again for all your help with this.

aeaversa commented:
Saw that some tests were failing the continuous integration checks and re-based to current master.

aeaversa commented:
Attaching screenshots...

![p-config](https://user-images.githubusercontent.com/36886905/52739181-59d54c80-2f9e-11e9-917a-6c768f5c5e1d.png)
![p-detail](https://user-images.githubusercontent.com/36886905/52739186-5b9f1000-2f9e-11e9-9e34-42332b5ab038.png)
![p-config-terminate](https://user-images.githubusercontent.com/36886905/52739191-5e9a0080-2f9e-11e9-9319-ded9450f6aec.png)

moranr commented:
@aeaversa, thanks for taking this on; it will be a great addition to NiFi! 

I would like to propose a slightly different way of presenting the info and guiding the user. It builds on some established design patterns for showing processor status and interactive conventions when stopping/terminating active threads.

I'm hoping this is straight forward referencing the attached mockup. This single image shows three dialog states, starting with Processor Details (left). This is as if the user double-clicked or selected 'view configuration' from a running processor's context menu. Here in the dialog there's a new section below the header that resembles the canvas's status bar. It's a familiar design, in a familiar location and it uses familiar icons along with the 'Run Status' labeling you'd see in the Summary table. The bulletin icon and styling is replicated as well. The new addition here is the 'stop & configure' action.

![view-config-dialogs](https://user-images.githubusercontent.com/14202495/53118471-286afc80-351b-11e9-8b70-5b5d883cf375.png)
Note: I used a processor that has an 'Advanced' configuration button as well, as that is another thing competing for real estate in this space.

The second dialog (center) is as if the user clicked the 'stop & configure' action and is now in the 'Configure Processor' dialog with editable fields. The processor status has changed to 'stopped', but the active thread continues. In this waiting state, the user can make changes, but the 'Apply' button is disabled and will remain so until the active thread completes. However, there is now a 'terminate' action available. This aligns with the current interactive convention seen from a processor's context menu (first you stop the processor, then the action is replaced with terminate if threads are running).

The third dialog (right) would be as if a thread completed on its own, or the user clicked the 'terminate' action. Here you can see the 'Apply' button is available allowing the user to proceed and continue as normal.

Interested in hearing your thoughts, along with @mcgilman 's as well.

aeaversa commented:
@moranr - Hey I really like this proposed change! I am on vacation thru next week, but if @mcgilman concurs, I'll refactor the current design to align with this specification when I get back and push out a fresh commit.Thanks again for the great feedback!

mcgilman commented:
@aeaversa @moranr I also like Rob's suggestions. They do a good job exposing the necessary operational details while limiting the number of places we need to do so. Each offers pretty subtle changes but allows for more streamlined user experience. Nice!

aeaversa commented:
@mcgilman @moranr - sounds good. I'll go ahead and refactor to the approved design when i get back. Thanks again for all the help with this.

aeaversa commented:
@mcgilman @moranr - Refactored all the requested changes in and pushed out an update to the branch. Attached are a few screen shots of how it all looks.  Please take a test drive and let me know your feedback whenever you guys get a chance. My thanks again for all your help with this.
![context](https://user-images.githubusercontent.com/36886905/54366109-7a003600-4646-11e9-9472-7867b0e5fead.png)
![details](https://user-images.githubusercontent.com/36886905/54366117-7cfb2680-4646-11e9-9114-09e30eea4fda.png)
![config1](https://user-images.githubusercontent.com/36886905/54366121-7ff61700-4646-11e9-8fbf-154c7aa91087.png)
![config2](https://user-images.githubusercontent.com/36886905/54366131-82f10780-4646-11e9-9980-7be4f36c9c37.png)

mcgilman commented:
@aeaversa Thanks for the updates. Good work here. A few more comments on the behavior. I haven't had a chance to dig in yet but want to share some initial thoughts. 

- I think the `Stop & Configure` menu item should be available when the component is stopped but there are active threads. This is most easily testable using `DebugFlow` and setting the `On Trigger Pause Time`. Should this menu item be visible under the same conditions as the `View configuration` menu item?
- Should the status bar be visible when it's unneeded?
- Quickly looking at the changes introduced into the jquery dialog component, I'm wondering if we should decouple the status bar content so that there's no dialog instance specific things in there. The fact that we're rendering component status, threads, and bulletins there are very specific to this use case and probably shouldn't be in the generic dialog component. What if we wanted to take this approach with other dialogs and other components that may report status differently (like active transmission for remote ports)? The fact that the button model is configurable is great.

@moranr Any thoughts? Icon for the `Stop & Configure` menu item look good?

moranr commented:
> I think the `Stop & Configure` menu item should be available when the component is stopped but there are active threads. This is most easily testable using `DebugFlow` and setting the `On Trigger Pause Time`. Should this menu item be visible under the same conditions as the `View configuration` menu item?

@mcgilman if I'm understanding correctly, this way you'll only ever see one option (_Configure_ or _View Configuration_ or _Stop & Configure_), right?

Also, if the above is true, we won't need to introduce another icon (fa-pencil-square). The gear icon can be used for all the options.

> Should the status bar be visible when it's unneeded?

You'd have access to bulletin details, so I'd say that is a good reason to keep it.

Other thoughts:
The icon + button text + background color in the status bar feels like a bit too much. I'm leaning more toward the text only option + subtle underline like in the mockups I put together. Thoughts?

aeaversa commented:
@mcgilman - Thanks again for the great feedback!

- As far as the "Stop & Configure" being persisted in the context menu, I do see where you are going, but think a clearer approach may be to simply modify the "View Configuration" to prompt as "Configure" when a processor is in a stopped state with active threads. I say this because the processor is already displaying the "Stopped" icon even when it has active threads and seeing a "Stop & Configure" button persisting probably would lead the user to think their prior "Stop & Configure" action was not successful. Thoughts?

- As far as the status bar being visible when its not needed, the feedback I've gathered from users has been pretty consistent in that when opened, the configuration dialog window tends to overlay the processor element on the graph and thus prevent the user from knowing the current state [ie- Started/Stopped/Invalid/Validating/Disabled]. The current implementation covers all states the processor could potentially achieve in the status bar and I can't see the harm in continuing to provide that state visualization. Thoughts?

- Lastly, I do see your point on not having the status bar as part of the jquery.modal component. I too wasn't a fan of putting it there, but I placed the logic there in order to avoid replicating code across the two dialog windows as it was a common point to both that I could leverage the button model logic already present. I will take a look at it and see if I can decouple that further to be more generic.

aeaversa commented:
@moranr - As far as the button text, the main concern I got from users was that when it was just plain text, users didn't realize it was actually clickable, so I tried to accentuate it. I'll mess with the style  to get it closer to your original vision but still keep the button effect. Thanks again for all your help with this.

moranr commented:
> @moranr - As far as the button text, the main concern I got from users was that when it was just plain text, users didn't realize it was actually clickable, so I tried to accentuate it. I'll mess with the style to get it closer to your original vision but still keep the button effect. Thanks again for all your help with this.

Overall the entire status bar will be new and I'd think easily noticeable, inviting users to explore more of what's there, but can understand some initial reaction you observed when testing. Things I tried to establish is a style unique to a clickable action (uppercase, same size and color font as a button) and including an underline suggestive of a link. I do think a background change on hover in this case would help/make sense since the bar is of similar (or maybe the same) height as a button and it is in close proximity to another element (the bulletin).

markobean commented:
@moranr I think it is flawed logic to rely on the status bar being "new" as an indicator the wording/button is a clickable object. It won't be "new" forever, and it should be just as obvious to a first-time user. I like the look and feel of @aeaversa approach.
The icon on the Stop & Configure in the status bar is the only thing I'm 50/50 on. If the icon stays, it has to remain as the fa-pencil-square to be consistent with the context menu since the gear already exists in this menu.

moranr commented:
> I think it is flawed logic to rely on the status bar being "new" as an indicator the wording/button is a clickable object. It won't be "new" forever, and it should be just as obvious to a first-time user.

@markobean I agree. I didn't mean to express the idea that something new would indicate something is clickable. My point was that visually the status bar, as a whole, is new. Someone seeing it for the first time would likely be curious about each part of it. Evaluating the appearance of the text alone is what I think indicates it's clickable. It is styled like other links in the app and similar to a button in font size/color/case treatment.

> The icon on the Stop & Configure in the status bar is the only thing I'm 50/50 on. If the icon stays, it has to remain as the fa-pencil-square to be consistent with the context menu since the gear already exists in this menu.

If I'm following the discussion by @mcgilman and @aeaversa correctly, we may only see a 'Stop & Configure' labeled action from the dialog with the new status bar. The context menu action will continue to say 'Configure' because it is stopped (even though active threads exist).

aeaversa commented:
> > @moranr - As far as the button text, the main concern I got from users was that when it was just plain text, users didn't realize it was actually clickable, so I tried to accentuate it. I'll mess with the style to get it closer to your original vision but still keep the button effect. Thanks again for all your help with this.
> 
> Overall the entire status bar will be new and I'd think easily noticeable, inviting users to explore more of what's there, but can understand some initial reaction you observed when testing. Things I tried to establish is a style unique to a clickable action (uppercase, same size and color font as a button) and including an underline suggestive of a link. I do think a background change on hover in this case would help/make sense since the bar is of similar (or maybe the same) height as a button and it is in close proximity to another element (the bulletin).
> 
> Another issue I have been trying to avoid adding to is introducing more icons. IMO NiFi has an awful lot already ;)

@moranr - Yeah, I agree. I will ensure the style is more subtle but still gets the idea across.

aeaversa commented:
> > I think it is flawed logic to rely on the status bar being "new" as an indicator the wording/button is a clickable object. It won't be "new" forever, and it should be just as obvious to a first-time user.
> 
> @markobean I agree. I didn't mean to express the idea that something new would indicate something is clickable. My point was that visually the status bar, as a whole, is new. Someone seeing it for the first time would likely be curious about each part of it. Evaluating the appearance of the text alone is what I think indicates it's clickable. It is styled like other links in the app and similar to a button in font size/color/case treatment.
> 
> > The icon on the Stop & Configure in the status bar is the only thing I'm 50/50 on. If the icon stays, it has to remain as the fa-pencil-square to be consistent with the context menu since the gear already exists in this menu.
> 
> If I'm following the discussion by @mcgilman and @aeaversa correctly, we may only see a 'Stop & Configure' labeled action from the dialog with the new status bar. The context menu action will continue to say 'Configure' because it is stopped (even though active threads exist).

@moranr - Actually, I think we are still keeping the "Stop & Configure" context menu entry. The main objective of the feature was seeking was to provide a more direct path for the user that allows them to edit a processor's properties, and having that direct option in the context menu i would say is super important. I think @mcgilman was just pointing out the fact that in the change the "View Configuration" menu option needs to be updated to "Configuration" when a processor is in a stopped state with active threads, since now the Processor Configuration modal will simply disable the Apply button and provide the user a Terminate button in the status bar should the user wish to apply their changes quickly. @mcgilman Let me know if I understood you correctly on that. I am not married to the pencil-square icon, but it does seem that all the context menu items all have an icon, so I'm open to suggestions. :-) Thanks to everyone for all the great feedback.

mcgilman commented:
I think the confusion here is stemming from the original JIRA and the comments on the motivation for the change above. The issue as described was that we have a Processor running someplace. Upon viewing the Processor Details, the user realizes they need to update some configuration and the process to do so is extremely cumbersome. I can see why @moranr was surprised by its addition of a new menu item. I was also thrown off by it a little bit which is why I wanted to make sure @moranr could review it. 

If we are considering a case where the Processor is running and the user wants to configure it (not from the Processor Details dialog) I would suggest we take a combined approach. The user would still select the Stop menu item from the context menu as they would today. From there we could immediately show the Configure menu item even if there are active threads since we won't activate the Apply button until those threads complete.

@aeaversa I can understand the code duplication concern with extracting the run state, threads, and bulletins out of the dialog. However, I think you should be able to create some utility function that Processors can utilize which can return some general type (maybe a jQuery instance) so we don't need that type of business logic within the dialog widget itself.

aeaversa commented:
Thanks @mcgilman . Yeah, my plan was to move the status bar code into its own jQuery based component that can be extrapolated onto any element. That way the code remains shared between both modals or whatever interface you want to consume the capability. As far as the context menu.... I too wasn't happy with the original icon, so I did redo it (see screen shots attached), I've also cleaned up the status bar button look and feel per @moranr 's feedback. Hopefully I've met the desired look and feel he was going for, but if not let me know.  I know the goal of this feature was attempting to provide an efficient navigational route to get the processor to a state where the user can make configuration changes, hopefully via a single click.  Adding the Stop & Configure option in the context menu does get the user there in one click, so are we sure we want to take that option away? My thanks again for all the great feedback on this.

![ss1](https://user-images.githubusercontent.com/36886905/55025380-39ee6b00-4fd7-11e9-9882-46af701516b7.png)
![ss2](https://user-images.githubusercontent.com/36886905/55025381-39ee6b00-4fd7-11e9-9b66-f4a9425ff9a9.png)
![ss2-mouse](https://user-images.githubusercontent.com/36886905/55025382-39ee6b00-4fd7-11e9-867e-f3b3088d41a0.png)
![ss3](https://user-images.githubusercontent.com/36886905/55025384-39ee6b00-4fd7-11e9-931d-e79bf6675e3b.png)
![ss4](https://user-images.githubusercontent.com/36886905/55025385-39ee6b00-4fd7-11e9-825f-430c36464e31.png)

moranr commented:
@aeaversa I like the new icon since it is a combination of the pre-existing ones.

> If we are considering a case where the Processor is running and the user wants to configure it (not from the Processor Details dialog) I would suggest we take a combined approach. The user would still select the Stop menu item from the context menu as they would today. From there we could immediately show the Configure menu item even if there are active threads since we won't activate the Apply button until those threads complete.

I agree with @mcgilman 's thoughts here. I think that would be a step in the right direction to alleviate some of the navigation deficiencies around configuration, but I don't think offering additional context menu options combining existing actions is part of the solution.

To further improvement in this area, I think having a separate discussion around changing the native behavior of NiFi could be worthwhile. By that I mean something like making 'configure' always available and if a component is running it is automatically stopped (for config changes to take place), then restarted (if it was previously running) once changes are applied.

aeaversa commented:
@moranr , @mcgilman - Ok, so for now I'll just go ahead and remove the Stop & Configure option in the context menu. As the new menu behavior,  I'll adjust the View Configuration menu option flip to Configure once a processor is in a stopped state with/without active threads. So essentially, the end user will have to click "Stop", then click "Configure" and if there are active threads, the "Apply" button will be unavailable until those thread resolve naturally or via a terminate call. Should have an updated commit for you guys later today.  Thanks again.

aeaversa commented:
@moranr , @mcgilman - Should have an update to the branch for you guys on Monday.

aeaversa commented:
@moranr , @mcgilman - Ok, the status bar has been decoupled into its own component and the context menu has been modified to display "Configure" versus "View Configuration" when a processor that is modifiable is semantically stopped, but retains active threads. Also I cleaned up the synchronization method with the canvas by adding a simple MutationObserver against the target canvas element. Please take a test drive and let me know if this works for you guys.  My thanks again for all your help with this.

aeaversa commented:
@moranr , @mcgilman - Some updated screen shots

![ss1](https://user-images.githubusercontent.com/36886905/55350174-76174500-5489-11e9-8630-ed8d4942b3b3.png)
![ss2](https://user-images.githubusercontent.com/36886905/55350175-76174500-5489-11e9-9851-215ab63cfe71.png)
![ss3](https://user-images.githubusercontent.com/36886905/55350176-76174500-5489-11e9-965f-1faf4a22be52.png)
![ss4](https://user-images.githubusercontent.com/36886905/55350177-76174500-5489-11e9-9b94-4cd536158c6f.png)
![ss5](https://user-images.githubusercontent.com/36886905/55350179-76174500-5489-11e9-979d-6741212b1d40.png)

moranr commented:
@aeaversa looks great based on the screenshot you included! I will try to pull this down soon and give it a go. I imagine @mcgilman will want to dig in a bit more as well.

mcgilman commented:
@aeaversa Sorry, I was away for the past week. I wanted to revisit this today but there appear to be some merge conflicts now. Can you rebase your branch? Thanks!

aeaversa commented:
@moranr , @mcgilman - Sorry for the delay on this. I went ahead and rebased to the latest master and resolved all the conflicts.  Look forward to your feedback.

markobean commented:
+1
I reviewed this from a usability point of view, and found no problems. This will help working with a flow enormously!
@moranr, @mcgilman, any additional comments?

moranr commented:
Sorry for the delay in getting back to look at this! 
+1 -- functionality looks good and it's working as I expected based on past discussion.

aeaversa commented:
@mcgilman -  I will look through these issues in the morning and try to get you out a fresh push tomorrow.  Thanks again for your help with this.

andrewmlim commented:
Tested the UI/UX and everything looks good - functionally works and intuitive. Also did some simple tests in a secured environment, making sure user who did not have permissions, did not see "Stop & Configure" and "Terminate".  Thanks for implementing this new feature @aeaversa !

aeaversa commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/jquery/statusbar/jquery.statusbar.js:
```diff
@@ -0,0 +1,287 @@
+            bar.statusbar('set',statusObj);
+
+            //create and store an observer
+            bar.data('observer',new MutationObserver(function(mutations){
```
@mcgilman - It was a function simply to allow the statusbar to re-query the object for the latest values when a mutation the DOM element is detected, as well as allowing other functionality to execute in the parent dialog when a mutation is observed. I agree its a little confusing so I refactored to simply pass in the processor id to query and a separate callback.

aeaversa commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/nf-processor-details.js:
```diff
@@ -296,6 +317,81 @@
+
+                    //Synchronize the status bar values to the canvas component
+                    var target = function(){
+                        return config.nfCanvasUtils.getSelectionById(processor.id)['datum']();
```
@mcgilman - Was initially troubleshooting an undefined issue an missed this one. Implemented change to directly invoke.

aeaversa commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/nf-processor-details.js:
```diff
@@ -296,6 +317,81 @@
+                            handler: {
+                                click: function() {
+                                    //Assign a callback when the ProcessorConfiguration window is ready
+                                    var cb = function(){
```
@mcgilman  - Refactored callback to remove duplicated code.

aeaversa commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-processor-configuration.js:
```diff
@@ -800,6 +826,10 @@
                             text: '#ffffff'
                         },
+                        disabled : function() {
+                            var selection = nfCanvasUtils.getSelectionById(processor.id);
```
@mcgilman - Refactored to remove duplicated selection.

aeaversa commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-processor-configuration.js:
```diff
@@ -885,8 +915,61 @@
+                                    text: '#004849'
+                                },
+                                disabled : function() {
+                                    var selection = nfCanvasUtils.getSelectionById(processor.id);
```
@mcgilman - Refactored to remove duplicated selection.

aeaversa commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-processor-configuration.js:
```diff
@@ -885,8 +915,61 @@
+
+                                        //execute the terminate call
+                                        $('#processor-configuration-status-bar').statusbar('hideButtons');
+                                        var selection = nfCanvasUtils.getSelectionById(processor.id);
```
@mcgilman - Refactored to remove duplicated selection.

aeaversa commented:
@mcgilman -  All requested changes have been pushed out to the branch. Take a test drive and let me know if there is anything else. Thanks again for all the help with this!

aeaversa commented:
> Tested the UI/UX and everything looks good - functionally works and intuitive. Also did some simple tests in a secured environment, making sure user who did not have permissions, did not see "Stop & Configure" and "Terminate". Thanks for implementing this new feature @aeaversa !

@andrewmlim  -  Thanks for the testing help!

mcgilman commented:
Thanks @aeaversa! This has been merged to master.
